Mathlib/Combinatorics/SimpleGraph/Turan.lean

```diff
@@ -91,4 +110,174 @@ theorem not_cliqueFree_of_isTuranMaximal (hn : r ≤ Fintype.card V) (hG : G.IsT
   exact hGab <| le_sup_right.trans_eq ((hG.le_iff_eq <| h.sup_edge _ _).1 le_sup_left).symm <|
     (edge_adj ..).2 ⟨Or.inl ⟨rfl, rfl⟩, hab⟩
 
+open Classical in
+lemma exists_IsTuranMaximal : ∃ H : SimpleGraph V, H.IsTuranMaximal r := by
+  let c := {H : SimpleGraph V | H.CliqueFree (r + 1)}
+  have cn : c.toFinset.Nonempty := ⟨⊥, by
+    simp only [Set.toFinset_setOf, mem_filter, mem_univ, true_and, c]
+    exact cliqueFree_bot (by omega)⟩
+  obtain ⟨S, Sm, Sl⟩ := exists_max_image c.toFinset (·.edgeFinset.card) cn
+  use S
+  rw [Set.mem_toFinset] at Sm
+  refine ⟨Sm, ?_⟩
+  intro I _ cf
+  by_cases Im : I ∈ c.toFinset
+  · convert Sl I Im
+  · rw [Set.mem_toFinset] at Im
+    contradiction
+
+end Defs
+
+section Forward
+
+variable {G} {s t u : V} (h : G.IsTuranMaximal r)
+
+namespace IsTuranMaximal
+
+/-- In a Turán-maximal graph, non-adjacent vertices have the same degree. -/
+lemma degree_eq_of_not_adj (hn : ¬G.Adj s t) : G.degree s = G.degree t := by
+  rw [IsTuranMaximal] at h; contrapose! h; intro cf
+  wlog hd : G.degree t < G.degree s generalizing G t s
+  · replace hd : G.degree s < G.degree t := lt_of_le_of_ne (le_of_not_lt hd) h
+    exact this (by rwa [adj_comm] at hn) hd.ne' cf hd
+  use G.replaceVertex s t, inferInstance, cf.replaceVertex s t
+  have := G.card_edgeFinset_replaceVertex_of_not_adj hn
+  omega
+
+/-- In a Turán-maximal graph, non-adjacency is transitive. -/
+lemma not_adj_transitive (hst : ¬G.Adj s t) (hsu : ¬G.Adj s u) : ¬G.Adj t u := by
```

Comment:
I confess I don't know this subject.
I first read this name as `not_(adj_transitive)` instead of `(not_adj)_transitive`.

Would it make sense to have an `abbrev` that allows you to write `G.notAdj s t`?

Reply:
There are other declarations in mathlib that contain `not_adj` in their names, like [`SimpleGraph.ne_of_adj_of_not_adj`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Combinatorics/SimpleGraph/Basic.html#SimpleGraph.ne_of_adj_of_not_adj) and [`SimpleGraph.Coloring.not_adj_of_mem_colorClass`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Combinatorics/SimpleGraph/Coloring.html#SimpleGraph.Coloring.not_adj_of_mem_colorClass). I guess the way out is to rename these theorems `transitive_of_not_adj` and `equivalence_of_not_adj`.

Reply:
…which I have now done.